File: app/ontime.py
from __future__ import annotations

from typing import Any, TypedDict
from urllib.parse import urlsplit, urlunsplit

import httpx


class OntimeError(RuntimeError):
    pass


class RundownData(TypedDict):
    title: str
    events: list[dict[str, Any]]
    custom_fields: list[str]


def build_rundown_url(base_url: str) -> str:
    """Append the data endpoint without losing an authenticated share-link token."""
    parsed = urlsplit(base_url.strip())
    path = f"{parsed.path.rstrip('/')}/data/rundowns/current/"
    return urlunsplit((parsed.scheme, parsed.netloc, path, parsed.query, ""))


def _unwrap_payload(data: Any) -> Any:
    if isinstance(data, dict) and "payload" in data:
        return data["payload"]
    return data


def extract_events(data: Any) -> list[dict[str, Any]]:
    """Tolerate the common Ontime current-rundown response shapes."""
    data = _unwrap_payload(data)
    if isinstance(data, list):
        return [item for item in data if isinstance(item, dict) and item.get("type") == "event"]
    if not isinstance(data, dict):
        raise OntimeError("Ontime returned an unexpected rundown format")

    entries = data.get("entries")
    if isinstance(entries, dict):
        order = data.get("flatOrder")
        if not isinstance(order, list):
            order = list(entries)
        return [
            entry
            for entry_id in order
            if isinstance((entry := entries.get(entry_id)), dict)
            and entry.get("type", "event").lower() == "event"
        ]

    for key in ("events", "rundown", "data"):
        value = data.get(key)
        if isinstance(value, list):
            return [item for item in value if isinstance(item, dict) and item.get("type", "event").lower() == "event"]
        if isinstance(value, dict):
            try:
                return extract_events(value)
            except OntimeError:
                pass
    raise OntimeError("No events were found in the current rundown")


def extract_rundown(data: Any) -> RundownData:
    """Extract printable rundown metadata while preserving event and field order."""
    unwrapped = _unwrap_payload(data)
    events = extract_events(unwrapped)
    title = unwrapped.get("title", "") if isinstance(unwrapped, dict) else ""
    custom_fields: list[str] = []
    for event in events:
        custom = event.get("custom")
        if not isinstance(custom, dict):
            continue
        for field in custom:
            if field not in custom_fields:
                custom_fields.append(field)
    return {"title": str(title), "events": events, "custom_fields": custom_fields}


async def fetch_current_rundown(
    base_url: str,
    auth_header: str | None = None,
    auth_value: str | None = None,
) -> RundownData:
    headers = {}
    if auth_header and auth_value:
        headers[auth_header] = auth_value
    url = build_rundown_url(base_url)
    try:
        async with httpx.AsyncClient(timeout=20, follow_redirects=True) as client:
            response = await client.get(url, headers=headers)
            if response.status_code == 401:
                raise OntimeError(
                    "Ontime rejected the connection. For a password-protected stage, paste an "
                    "authenticated Companion share link from Ontime's Sharing and reporting settings."
                )
            response.raise_for_status()
            if "application/json" not in response.headers.get("content-type", ""):
                raise OntimeError("Ontime returned a non-JSON response; check the stage URL and login requirements")
            return extract_rundown(response.json())
    except OntimeError:
        raise
    except httpx.HTTPError as exc:
        raise OntimeError(f"Could not read the Ontime rundown: {exc}") from exc

File: netlify/functions/_shared/ontime.mjs
// Copyright 2026 Steve Weiss
// SPDX-License-Identifier: 0BSD

export class OntimeError extends Error {}

export const ONTIME_CLOUD_HOST = "cloud.getontime.no";
export const MAX_ONTIME_RESPONSE_BYTES = 5 * 1024 * 1024;

export function validateOntimeUrl(baseUrl) {
  const rawUrl = String(baseUrl ?? "").trim();
  if (!rawUrl || rawUrl.length > 4096) {
    throw new OntimeError("Enter a valid Ontime Cloud stage or Companion share URL");
  }

  let parsed;
  try {
    parsed = new URL(rawUrl);
  } catch {
    throw new OntimeError("Enter a valid Ontime Cloud stage or Companion share URL");
  }

  const hostname = parsed.hostname.toLowerCase().replace(/\.$/, "");
  if (parsed.protocol !== "https:" || hostname !== ONTIME_CLOUD_HOST) {
    throw new OntimeError(`Only HTTPS Ontime Cloud URLs from ${ONTIME_CLOUD_HOST} are supported`);
  }
  if (parsed.username || parsed.password || (parsed.port && parsed.port !== "443")) {
    throw new OntimeError("Ontime Cloud URLs cannot contain credentials or a custom port");
  }
  if (parsed.hash) {
    throw new OntimeError("Remove the fragment from the Ontime Cloud URL and try again");
  }
  if (!parsed.pathname.replaceAll("/", "")) {
    throw new OntimeError("The Ontime Cloud URL must include a stage identifier");
  }
}
export function buildDataUrl(baseUrl, resource) {
  const parsed = new URL(String(baseUrl).trim());
  parsed.pathname = `${parsed.pathname.replace(/\/+$/, "")}/data/${String(resource).replace(/^\/+|\/+$/g, "")}/`;
  parsed.hash = "";
  return parsed.toString();
}

export function buildRundownUrl(baseUrl) {
  return buildDataUrl(baseUrl, "rundowns/current");
}

export function buildProjectUrl(baseUrl) {
  return buildDataUrl(baseUrl, "project");
}

function unwrapPayload(data) {
  return data && typeof data === "object" && !Array.isArray(data) && "payload" in data
    ? data.payload
    : data;
}

export function extractEvents(data) {
  const unwrapped = unwrapPayload(data);
  if (Array.isArray(unwrapped)) {
    return unwrapped.filter((item) => item && typeof item === "object" && item.type === "event");
  }
  if (!unwrapped || typeof unwrapped !== "object") {
    throw new OntimeError("Ontime returned an unexpected rundown format");
  }

  if (unwrapped.entries && typeof unwrapped.entries === "object" && !Array.isArray(unwrapped.entries)) {
    const order = Array.isArray(unwrapped.flatOrder)
      ? unwrapped.flatOrder
      : Object.keys(unwrapped.entries);
    return order
      .map((entryId) => unwrapped.entries[entryId])
      .filter(
        (entry) =>
          entry &&
          typeof entry === "object" &&
          String(entry.type ?? "event").toLowerCase() === "event",
      );
  }

  for (const key of ["events", "rundown", "data"]) {
    const value = unwrapped[key];
    if (Array.isArray(value)) {
      return value.filter(
        (item) =>
          item && typeof item === "object" && String(item.type ?? "event").toLowerCase() === "event",
      );
    }
    if (value && typeof value === "object") {
      try {
        return extractEvents(value);
      } catch (error) {
        if (!(error instanceof OntimeError)) throw error;
      }
    }
  }
  throw new OntimeError("No events were found in the current rundown");
}

export function extractProjectTitle(data) {
  const unwrapped = unwrapPayload(data);
  if (!unwrapped || typeof unwrapped !== "object" || Array.isArray(unwrapped)) return "";
  return String(unwrapped.title ?? "").trim();
}

export function extractRundown(data, projectData = null) {
  const unwrapped = unwrapPayload(data);
  const events = extractEvents(unwrapped);
  const rundownTitle =
    unwrapped && typeof unwrapped === "object" && !Array.isArray(unwrapped)
      ? String(unwrapped.title ?? "").trim()
      : "";
  const title = extractProjectTitle(projectData) || rundownTitle;
  const customFields = [];
  for (const event of events) {
    const custom = event.custom;
    if (!custom || typeof custom !== "object" || Array.isArray(custom)) continue;
    for (const field of Object.keys(custom)) {
      if (!customFields.includes(field)) customFields.push(field);
    }
  }
  return { title, rundown_title: rundownTitle, events, custom_fields: customFields };
}

async function readJsonResponse(response, tooLargeMessage) {
  const declaredSize = Number(response.headers.get("content-length") || 0);
  if (declaredSize > MAX_ONTIME_RESPONSE_BYTES) throw new OntimeError(tooLargeMessage);

  const reader = response.body?.getReader();
  if (!reader) return response.json();
  const chunks = [];
  let total = 0;
  while (true) {
    const { value, done } = await reader.read();
    if (done) break;
    total += value.byteLength;
    if (total > MAX_ONTIME_RESPONSE_BYTES) {
      await reader.cancel();
      throw new OntimeError(tooLargeMessage);
    }
    chunks.push(value);
  }
  const bytes = new Uint8Array(total);
  let offset = 0;
  for (const chunk of chunks) {
    bytes.set(chunk, offset);
    offset += chunk.byteLength;
  }
  return JSON.parse(new TextDecoder().decode(bytes));
}

export async function fetchCurrentRundown(baseUrl, options = {}) {
  validateOntimeUrl(baseUrl);
  const headers = {};
  if (options.authHeader && options.authValue) headers[options.authHeader] = options.authValue;
  const requestOptions = {
    headers,
    redirect: "manual",
    signal: AbortSignal.timeout(20_000),
  };

  try {
    const [rundownResult, projectResult] = await Promise.allSettled([
      fetch(buildRundownUrl(baseUrl), requestOptions),
      fetch(buildProjectUrl(baseUrl), requestOptions),
    ]);
    if (rundownResult.status === "rejected") throw rundownResult.reason;
    const rundownResponse = rundownResult.value;
    if (rundownResponse.status >= 300 && rundownResponse.status < 400) {
      throw new OntimeError(
        "Ontime redirected the request unexpectedly; check the stage or Companion share URL",
      );
    }
    if ([401, 403].includes(rundownResponse.status)) {
      throw new OntimeError(
        "Ontime rejected the connection. For a password-protected stage, paste an authenticated Companion share link from Ontime's Sharing and reporting settings.",
      );
    }
    if (!rundownResponse.ok) {
      throw new OntimeError(`Ontime returned HTTP ${rundownResponse.status}`);
    }
    if (!rundownResponse.headers.get("content-type")?.toLowerCase().includes("application/json")) {
      throw new OntimeError(
        "Ontime returned a non-JSON response; check the stage URL and login requirements",
      );
    }
    const rundownData = await readJsonResponse(
      rundownResponse,
      "The Ontime rundown response was too large to process safely",
    );

    let projectData = null;
    if (projectResult.status === "fulfilled") {
      const projectResponse = projectResult.value;
      if (
        projectResponse.ok &&
        projectResponse.headers.get("content-type")?.toLowerCase().includes("application/json")
      ) {
        try {
          projectData = await readJsonResponse(
            projectResponse,
            "The Ontime project response was too large to process safely",
          );
        } catch {
          projectData = null;
        }
      }
    }
    return extractRundown(rundownData, projectData);
  } catch (error) {
    if (error instanceof OntimeError) throw error;
    throw new OntimeError("Could not read the Ontime rundown; check the stage URL and try again");
  }
}
